File: stress-tests/sdk-runners/node/src/runner.ts
import { createServer, type IncomingMessage, type Server, type ServerResponse } from "node:http";
import { createConfig, NodeRequestContextResolver, ToToggleClient } from "totoggle-node";

const DEFAULT_BIND_ADDRESS = "127.0.0.1";
const DEFAULT_PORT = 19092;
const DEFAULT_TIMEOUT_MS = 10_000;
const MAX_REQUEST_BYTES = 16 * 1024;

export interface RunnerConfig {
  readonly serverUrl: string;
  readonly secretKey: string;
  readonly httpTimeoutMs: number;
  readonly bindAddress?: string;
  readonly port?: number;
  readonly allowNonLoopbackTarget?: boolean;
}

export interface Runner {
  readonly handle: (request: IncomingMessage, response: ServerResponse) => Promise<void>;
  close(): void;
}

interface DomainContext {
  readonly userId?: string;
  readonly rolloutKey?: string;
  readonly cohort?: string;
  readonly attributes?: Readonly<Record<string, string>>;
}

interface EvaluationRequest {
  readonly path: string;
  readonly context: DomainContext;
}

/**
 * Starts the production Node SDK and wraps it in a local stress-only HTTP surface. Domain values
 * are request-local through NodeRequestContextResolver; network values remain derived from the
 * actual loopback connection plus trusted Forwarded/CF-IPCountry headers.
 */
export async function createRunner(input: RunnerConfig): Promise<Runner> {
  const config = normalizeAndValidate(input);
  const contexts = new WeakMap<IncomingMessage, DomainContext>();
  const resolver = new NodeRequestContextResolver({
    trustedProxyAddresses: ["127.0.0.1", "::1"],
    values: (request) => contexts.get(request),
  });
  const client = new ToToggleClient(createConfig(
    "totoggle-stress-node-runner",
    config.serverUrl,
    config.secretKey,
    { httpTimeoutMs: config.httpTimeoutMs, refreshIntervalMs: 5_000, contextResolver: resolver },
  ));
  await client.start();
  if (!client.isHealthy()) {
    client.shutdown();
    throw new Error("SDK catalog was not available at startup");
  }

  let closed = false;
  return {
    handle: async (request, response) => {
      try {
        await handleEvaluation(request, response, client, resolver, contexts, () => closed);
      } catch {
        // Neither errors nor request data are logged: callers receive a stable, non-sensitive
        // response even if malformed input or an application resolver misbehaves.
        if (!response.headersSent) writeRejected(response, 500);
      }
    },
    close: () => {
      if (closed) return;
      closed = true;
      client.shutdown();
    },
  };
}

export function createSidecarServer(runner: Runner): Server {
  const server = createServer((request, response) => { void runner.handle(request, response); });
  server.requestTimeout = 10_000;
  server.headersTimeout = 5_000;
  server.keepAliveTimeout = 30_000;
  return server;
}

async function handleEvaluation(
  request: IncomingMessage,
  response: ServerResponse,
  client: ToToggleClient,
  resolver: NodeRequestContextResolver,
  contexts: WeakMap<IncomingMessage, DomainContext>,
  isClosed: () => boolean,
): Promise<void> {
  if (request.url === "/health") {
    if (request.method !== "GET") return writeRejected(response, 405);
    if (isClosed() || !client.isHealthy()) return writeRejected(response, 503);
    return writeJson(response, 200, { status: "ok" });
  }
  if (request.url !== "/evaluate") return writeRejected(response, 404);
  if (request.method !== "POST") return writeRejected(response, 405);
  if (!request.headers["content-type"]?.toLowerCase().startsWith("application/json")) return writeRejected(response, 415);
  const input = await decodeRequest(request);
  if (!input) return writeRejected(response, 400);
  if (isClosed() || !client.isHealthy()) return writeRejected(response, 503);

  contexts.set(request, input.context);
  try {
    const active = resolver.run(request, () => client.isActive(input.path));
    writeJson(response, 200, { active });
  } finally {
    contexts.delete(request);
  }
}

async function decodeRequest(request: IncomingMessage): Promise<EvaluationRequest | undefined> {
  const contentLength = Number(request.headers["content-length"]);
  if (Number.isFinite(contentLength) && contentLength > MAX_REQUEST_BYTES) return undefined;
  const chunks: Buffer[] = [];
  let length = 0;
  for await (const chunk of request) {
    const buffer = Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk);
    length += buffer.length;
    if (length > MAX_REQUEST_BYTES) return undefined;
    chunks.push(buffer);
  }
  try {
    return validateRequest(JSON.parse(Buffer.concat(chunks).toString("utf8")));
  } catch {
    return undefined;
  }
}

function validateRequest(value: unknown): EvaluationRequest | undefined {
  if (!isPlainObject(value) || !hasOnlyKeys(value, ["path", "context"]) || typeof value.path !== "string" || value.path.trim() === "" || value.path.length > 512) return undefined;
  const context = validateContext(value.context);
  return context === undefined ? undefined : { path: value.path, context };
}

function validateContext(value: unknown): DomainContext | undefined {
  if (!isPlainObject(value) || !hasOnlyKeys(value, ["userId", "rolloutKey", "cohort", "attributes"])) return undefined;
  const strings = [value.userId, value.rolloutKey, value.cohort];
  if (strings.some((field) => field !== undefined && (typeof field !== "string" || field.length > 1_024))) return undefined;
  const attributes = validateAttributes(value.attributes);
  if (attributes === undefined && value.attributes !== undefined) return undefined;
  return { userId: value.userId as string | undefined, rolloutKey: value.rolloutKey as string | undefined, cohort: value.cohort as string | undefined, attributes };
}

function validateAttributes(value: unknown): Readonly<Record<string, string>> | undefined {
  if (value === undefined) return undefined;
  if (!isPlainObject(value)) return undefined;
  const entries = Object.entries(value);
  if (entries.length > 32 || entries.some(([name, field]) => name.trim() === "" || name.length > 128 || typeof field !== "string" || field.length > 1_024)) return undefined;
  return Object.freeze(Object.fromEntries(entries as [string, string][]));
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value) && Object.getPrototypeOf(value) === Object.prototype;
}

function hasOnlyKeys(value: Record<string, unknown>, keys: readonly string[]): boolean {
  return Object.keys(value).every((key) => keys.includes(key));
}

function writeRejected(response: ServerResponse, status: number): void {
  writeJson(response, status, { error: "request rejected" });
}

function writeJson(response: ServerResponse, status: number, body: object): void {
  response.writeHead(status, { "content-type": "application/json", "cache-control": "no-store" });
  response.end(JSON.stringify(body));
}

export function loadConfigFromEnv(environment: NodeJS.ProcessEnv): Required<RunnerConfig> {
  const serverUrl = environment.STRESS_SERVER_URL?.trim();
  const secretKey = environment.STRESS_SECRET_KEY?.trim();
  if (!serverUrl || !secretKey) throw new Error("STRESS_SERVER_URL and STRESS_SECRET_KEY are required");
  const parsedPort = parseNumber(environment.STRESS_NODE_PORT, DEFAULT_PORT, "STRESS_NODE_PORT", 1, 65_535);
  const timeout = parseNumber(environment.STRESS_HTTP_TIMEOUT_MS, DEFAULT_TIMEOUT_MS, "STRESS_HTTP_TIMEOUT_MS", 1, 120_000);
  return normalizeAndValidate({
    serverUrl,
    secretKey,
    httpTimeoutMs: timeout,
    bindAddress: environment.STRESS_NODE_BIND?.trim() || DEFAULT_BIND_ADDRESS,
    port: parsedPort,
    allowNonLoopbackTarget: environment.ALLOW_NON_LOOPBACK_STRESS_TARGETS === "yes",
  });
}

function normalizeAndValidate(config: RunnerConfig): Required<RunnerConfig> {
  let url: URL;
  try {
    url = new URL(config.serverUrl);
  } catch {
    throw new Error("STRESS_SERVER_URL must be an absolute HTTP(S) URL without credentials");
  }
  if (!(["http:", "https:"].includes(url.protocol)) || url.username || url.password || !url.hostname || url.search || url.hash) throw new Error("STRESS_SERVER_URL must be an absolute HTTP(S) URL without credentials");
  if (!config.allowNonLoopbackTarget && !isLoopbackHost(url.hostname)) throw new Error("STRESS_SERVER_URL must not target a non-loopback host unless ALLOW_NON_LOOPBACK_STRESS_TARGETS=yes");
  if (!config.secretKey.startsWith("sk_")) throw new Error("STRESS_SECRET_KEY must be a valid secret key");
  if (!Number.isInteger(config.httpTimeoutMs) || config.httpTimeoutMs < 1 || config.httpTimeoutMs > 120_000) throw new Error("HTTP timeout must be between 1 and 120000");
  const bindAddress = config.bindAddress ?? DEFAULT_BIND_ADDRESS;
  if (bindAddress !== "127.0.0.1" && bindAddress !== "::1") throw new Error("stress runner bind address must be loopback");
  const port = config.port ?? DEFAULT_PORT;
  if (!Number.isInteger(port) || port < 1 || port > 65_535) throw new Error("stress runner port must be between 1 and 65535");
  return { serverUrl: config.serverUrl, secretKey: config.secretKey, httpTimeoutMs: config.httpTimeoutMs, bindAddress, port, allowNonLoopbackTarget: config.allowNonLoopbackTarget ?? false };
}

function isLoopbackHost(host: string): boolean {
  return host === "localhost" || host === "127.0.0.1" || host === "[::1]" || host === "::1";
}

function parseNumber(value: string | undefined, defaultValue: number, name: string, minimum: number, maximum: number): number {
  if (value === undefined || value.trim() === "") return defaultValue;
  const parsed = Number(value);
  if (!Number.isInteger(parsed) || parsed < minimum || parsed > maximum) throw new Error(`${name} must be between ${minimum} and ${maximum}`);
  return parsed;
}

File: stress-tests/sdk-runners/go/runner.go
package main

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/netip"
	"net/url"
	"os"
	"strconv"
	"strings"
	"sync"
	"time"

	totoggle "github.com/manorfm/toToggles/totoggle_go"
	"github.com/manorfm/toToggles/totoggle_go/httpcontext"
)

const (
	defaultBindAddress = "127.0.0.1"
	defaultPort        = 19091
	maxRequestBytes    = 16 << 10
)

// RunnerConfig configures a local-only stress sidecar. ServerURL and SecretKey are deliberately
// explicit: a stress runner must never silently target an arbitrary environment.
type RunnerConfig struct {
	ServerURL              string
	SecretKey              string
	HTTPTimeout            time.Duration
	BindAddress            string
	Port                   int
	AllowNonLoopbackTarget bool
}

// Runner evaluates flags through the production Go SDK. It does not reproduce SDK evaluation
// logic, so stress scenarios exercise catalog synchronization and request-local context exactly
// as an HTTP application would.
type Runner struct {
	client    *totoggle.Client
	resolver  *httpcontext.Resolver
	closeOnce sync.Once
}

type scenarioContext struct {
	UserID     string            `json:"userId"`
	RolloutKey string            `json:"rolloutKey"`
	Cohort     string            `json:"cohort"`
	Attributes map[string]string `json:"attributes"`
}

type evaluationRequest struct {
	Path    string          `json:"path"`
	Context scenarioContext `json:"context"`
}

type scenarioContextKey struct{}

// NewRunner starts an SDK client with a loopback-trusted HTTP context adapter. Network context
// is still derived exclusively from Forwarded/CF-IPCountry headers on the incoming request;
// request JSON contains only application-owned values.
func NewRunner(config RunnerConfig) (*Runner, error) {
	config = normalizedRunnerConfig(config)
	if err := validateRunnerConfig(config); err != nil {
		return nil, err
	}
	resolver := httpcontext.New(httpcontext.Options{
		TrustedProxyAddresses: []string{"127.0.0.1", "::1"},
		ApplicationValues: func(request *http.Request) httpcontext.ApplicationValues {
			input, _ := request.Context().Value(scenarioContextKey{}).(scenarioContext)
			return httpcontext.ApplicationValues{
				UserID:     input.UserID,
				RolloutKey: input.RolloutKey,
				Cohort:     input.Cohort,
				Attributes: input.Attributes,
			}
		},
	})
	clientConfig, err := totoggle.NewConfig(
		"totoggle-stress-go-runner",
		config.ServerURL,
		config.SecretKey,
		totoggle.WithHTTPTimeout(config.HTTPTimeout),
		totoggle.WithRefreshInterval(5*time.Second),
		totoggle.WithToggleContextResolver(resolver),
	)
	if err != nil {
		return nil, fmt.Errorf("invalid SDK configuration: %w", err)
	}
	client := totoggle.New(clientConfig)
	if err := client.Start(context.Background()); err != nil || !client.IsHealthy() {
		client.Shutdown()
		return nil, errors.New("SDK catalog was not available at startup")
	}
	return &Runner{client: client, resolver: resolver}, nil
}

func normalizedRunnerConfig(config RunnerConfig) RunnerConfig {
	if config.BindAddress == "" {
		config.BindAddress = defaultBindAddress
	}
	if config.Port == 0 {
		config.Port = defaultPort
	}
	return config
}

func validateRunnerConfig(config RunnerConfig) error {
	parsedURL, err := url.ParseRequestURI(config.ServerURL)
	if err != nil || (parsedURL.Scheme != "http" && parsedURL.Scheme != "https") || parsedURL.Host == "" || parsedURL.User != nil || parsedURL.RawQuery != "" || parsedURL.Fragment != "" {
		return errors.New("STRESS_SERVER_URL must be an absolute HTTP(S) URL without credentials")
	}
	if !config.AllowNonLoopbackTarget && !isLoopbackHost(parsedURL.Hostname()) {
		return errors.New("STRESS_SERVER_URL must not target a non-loopback host unless ALLOW_NON_LOOPBACK_STRESS_TARGETS=yes")
	}
	if !strings.HasPrefix(config.SecretKey, "sk_") {
		return errors.New("STRESS_SECRET_KEY must be a valid secret key")
	}
	if config.HTTPTimeout <= 0 {
		return errors.New("HTTP timeout must be positive")
	}
	if config.BindAddress == "" {
		config.BindAddress = defaultBindAddress
	}
	address, err := netip.ParseAddr(config.BindAddress)
	if err != nil || !address.IsLoopback() {
		return errors.New("stress runner bind address must be loopback")
	}
	if config.Port < 1 || config.Port > 65535 {
		return errors.New("stress runner port must be between 1 and 65535")
	}
	return nil
}

func isLoopbackHost(host string) bool {
	return strings.EqualFold(host, "localhost") || host == "127.0.0.1" || host == "::1"
}

// Handler returns the local sidecar API. POST /evaluate accepts only domain context:
// {"path":"payments.card","context":{"userId":"...","rolloutKey":"...","cohort":"...","attributes":{"plan":"pro"}}}.
// IP and country are intentionally omitted from this JSON and come from trusted proxy headers.
func (r *Runner) Handler() http.Handler {
	return http.HandlerFunc(r.serveHTTP)
}

func (r *Runner) serveHTTP(writer http.ResponseWriter, request *http.Request) {
	if request.URL.Path == "/health" {
		if request.Method != http.MethodGet {
			writeStatus(writer, http.StatusMethodNotAllowed)
			return
		}
		if !r.client.IsHealthy() {
			writeStatus(writer, http.StatusServiceUnavailable)
			return
		}
		writer.Header().Set("Content-Type", "application/json")
		writer.WriteHeader(http.StatusOK)
		_, _ = writer.Write([]byte(`{"status":"ok"}`))
		return
	}
	if request.URL.Path != "/evaluate" {
		writeStatus(writer, http.StatusNotFound)
		return
	}
	if request.Method != http.MethodPost {
		writeStatus(writer, http.StatusMethodNotAllowed)
		return
	}
	if !strings.HasPrefix(request.Header.Get("Content-Type"), "application/json") {
		writeStatus(writer, http.StatusUnsupportedMediaType)
		return
	}
	input, ok := decodeEvaluationRequest(request.Body)
	if !ok {
		writeStatus(writer, http.StatusBadRequest)
		return
	}
	if !r.client.IsHealthy() {
		writeStatus(writer, http.StatusServiceUnavailable)
		return
	}

	request = request.WithContext(context.WithValue(request.Context(), scenarioContextKey{}, input.Context))
	var active bool
	r.resolver.Middleware(http.HandlerFunc(func(_ http.ResponseWriter, contextualRequest *http.Request) {
		active = r.client.IsActiveContext(contextualRequest.Context(), input.Path)
	})).ServeHTTP(writer, request)
	writer.Header().Set("Content-Type", "application/json")
	writer.WriteHeader(http.StatusOK)
	_, _ = writer.Write([]byte(`{"active":` + strconv.FormatBool(active) + `}`))
}

func decodeEvaluationRequest(body io.ReadCloser) (evaluationRequest, bool) {
	defer body.Close()
	decoder := json.NewDecoder(io.LimitReader(body, maxRequestBytes+1))
	decoder.DisallowUnknownFields()
	var input evaluationRequest
	if decoder.Decode(&input) != nil || decoder.Decode(&struct{}{}) != io.EOF || !validInput(input) {
		return evaluationRequest{}, false
	}
	return input, true
}

func validInput(input evaluationRequest) bool {
	if strings.TrimSpace(input.Path) == "" || len(input.Path) > 512 || len(input.Context.Attributes) > 32 {
		return false
	}
	for _, value := range []string{input.Context.UserID, input.Context.RolloutKey, input.Context.Cohort} {
		if len(value) > 1024 {
			return false
		}
	}
	for name, value := range input.Context.Attributes {
		if strings.TrimSpace(name) == "" || len(name) > 128 || len(value) > 1024 {
			return false
		}
	}
	return true
}

func writeStatus(writer http.ResponseWriter, status int) {
	writer.Header().Set("Content-Type", "application/json")
	writer.WriteHeader(status)
	_, _ = writer.Write([]byte(`{"error":"request rejected"}`))
}

// Close releases the SDK refresh loop. The context is accepted so command shutdown can use the
// same lifecycle shape as http.Server.Shutdown; the SDK shutdown itself is synchronous.
func (r *Runner) Close(_ context.Context) {
	r.closeOnce.Do(r.client.Shutdown)
}

// LoadConfigFromEnv reads explicit runner configuration without ever including secrets in errors.
func LoadConfigFromEnv() (RunnerConfig, error) {
	serverURL := strings.TrimSpace(os.Getenv("STRESS_SERVER_URL"))
	secretKey := strings.TrimSpace(os.Getenv("STRESS_SECRET_KEY"))
	if serverURL == "" || secretKey == "" {
		return RunnerConfig{}, errors.New("STRESS_SERVER_URL and STRESS_SECRET_KEY are required")
	}
	port := defaultPort
	if raw := strings.TrimSpace(os.Getenv("STRESS_GO_PORT")); raw != "" {
		parsed, err := strconv.Atoi(raw)
		if err != nil {
			return RunnerConfig{}, errors.New("STRESS_GO_PORT must be a number")
		}
		port = parsed
	}
	timeout := 10 * time.Second
	if raw := strings.TrimSpace(os.Getenv("STRESS_HTTP_TIMEOUT_MS")); raw != "" {
		milliseconds, err := strconv.Atoi(raw)
		if err != nil || milliseconds < 1 || milliseconds > 120000 {
			return RunnerConfig{}, errors.New("STRESS_HTTP_TIMEOUT_MS must be between 1 and 120000")
		}
		timeout = time.Duration(milliseconds) * time.Millisecond
	}
	config := RunnerConfig{
		ServerURL:              serverURL,
		SecretKey:              secretKey,
		HTTPTimeout:            timeout,
		BindAddress:            strings.TrimSpace(os.Getenv("STRESS_GO_BIND")),
		Port:                   port,
		AllowNonLoopbackTarget: os.Getenv("ALLOW_NON_LOOPBACK_STRESS_TARGETS") == "yes",
	}
	if config.BindAddress == "" {
		config.BindAddress = defaultBindAddress
	}
	if err := validateRunnerConfig(config); err != nil {
		return RunnerConfig{}, err
	}
	return config, nil
}

func listenAddress(config RunnerConfig) string {
	return net.JoinHostPort(config.BindAddress, strconv.Itoa(config.Port))
}
